File: src/analyzer/work-leverage.ts
/**
 * Work Leverage Metrics
 *
 * Measures how effectively later agents build on earlier agents' work.
 * All metrics are computed from git diffs — no reference to coordination tools used.
 *
 * Key principle: rework that improves quality (tech debt paydown) is scored
 * differently from uninformed rework (agent didn't know what predecessor did).
 */

import type { AgentTranscript, FileChange } from '../types/transcript.js';

/**
 * Result of work leverage analysis across ordered sessions.
 */
export interface WorkLeverageResult {
  /** Per-session rework ratios: fraction of session N's additions deleted by N+1 */
  reworkRatios: number[];
  /** Average rework ratio across all session pairs */
  avgReworkRatio: number;
  /** Per-session line survival rates: fraction of session N's additions surviving to final state */
  lineSurvivalRates: number[];
  /** Average line survival rate */
  avgLineSurvivalRate: number;
  /** Per-session continuation indices: fraction of B's additions that reference A's new symbols */
  continuationIndices: number[];
  /** Average continuation index */
  avgContinuationIndex: number;
}

/**
 * Parse a unified diff to extract added and removed line content (without +/- prefix).
 * Filters to meaningful code lines, ignoring blank lines and diff headers.
 */
function parseAddedLines(diff: string): string[] {
  return diff
    .split('\n')
    .filter((line) => line.startsWith('+') && !line.startsWith('+++'))
    .map((line) => line.slice(1).trim())
    .filter((line) => line.length > 0);
}

function parseRemovedLines(diff: string): string[] {
  return diff
    .split('\n')
    .filter((line) => line.startsWith('-') && !line.startsWith('---'))
    .map((line) => line.slice(1).trim())
    .filter((line) => line.length > 0);
}

/**
 * Filter to source code files only (excludes docs, coordination artifacts).
 */
function filterSourceFiles(fileChanges: FileChange[]): FileChange[] {
  return fileChanges.filter(
    (c) => (c.path.startsWith('src/') || c.path.startsWith('tests/')) && c.path.endsWith('.ts'),
  );
}

/**
 * Compute rework ratio between two consecutive sessions.
 *
 * reworkRatio = (lines A added that B removed) / (lines A added)
 *
 * A low ratio means B built on A's work. A high ratio means B rewrote A's work.
 */
export function computeReworkRatio(
  sessionA: AgentTranscript,
  sessionB: AgentTranscript,
): number {
  const aSourceFiles = filterSourceFiles(sessionA.fileChanges);
  const bSourceFiles = filterSourceFiles(sessionB.fileChanges);

  const aDiffs = aSourceFiles
    .map((c) => c.diff)
    .filter((d): d is string => d !== undefined)
    .join('\n');
  const bDiffs = bSourceFiles
    .map((c) => c.diff)
    .filter((d): d is string => d !== undefined)
    .join('\n');

  const aAdded = new Set(parseAddedLines(aDiffs));
  const bRemoved = new Set(parseRemovedLines(bDiffs));

  if (aAdded.size === 0) return 0;

  // Count how many of A's added lines were removed by B
  let overlapCount = 0;
  for (const line of aAdded) {
    if (bRemoved.has(line)) {
      overlapCount++;
    }
  }

  return overlapCount / aAdded.size;
}

/**
 * Compute line survival rate for a session: what fraction of its additions
 * survive through all subsequent sessions to the final state.
 *
 * lineSurvival = (session N lines NOT removed by any later session) / (session N total additions)
 */
export function computeLineSurvivalRate(
  session: AgentTranscript,
  laterSessions: AgentTranscript[],
): number {
  const sourceFiles = filterSourceFiles(session.fileChanges);
  const diffs = sourceFiles
    .map((c) => c.diff)
    .filter((d): d is string => d !== undefined)
    .join('\n');

  const added = new Set(parseAddedLines(diffs));
  if (added.size === 0) return 1; // No additions = nothing to survive

  // Collect all lines removed by any later session
  const allLaterRemoved = new Set<string>();
  for (const later of laterSessions) {
    const laterSourceFiles = filterSourceFiles(later.fileChanges);
    const laterDiffs = laterSourceFiles
      .map((c) => c.diff)
      .filter((d): d is string => d !== undefined)
      .join('\n');
    for (const line of parseRemovedLines(laterDiffs)) {
      allLaterRemoved.add(line);
    }
  }

  let survivingCount = 0;
  for (const line of added) {
    if (!allLaterRemoved.has(line)) {
      survivingCount++;
    }
  }

  return survivingCount / added.size;
}

/**
 * Compute continuation index: did session B extend A's code structure
 * (importing A's new symbols, calling A's functions) or create parallel implementations?
 *
 * continuationIndex = (B's additions referencing A's new symbols) / (B's total additions)
 *
 * A's "new symbols" are identifiers extracted from A's added export/class/function/interface declarations.
 */
export function computeContinuationIndex(
  sessionA: AgentTranscript,
  sessionB: AgentTranscript,
): number {
  const aSourceFiles = filterSourceFiles(sessionA.fileChanges);
  const bSourceFiles = filterSourceFiles(sessionB.fileChanges);

  const aDiffs = aSourceFiles
    .map((c) => c.diff)
    .filter((d): d is string => d !== undefined)
    .join('\n');
  const bDiffs = bSourceFiles
    .map((c) => c.diff)
    .filter((d): d is string => d !== undefined)
    .join('\n');

  // Extract symbol names from A's added lines (exports, classes, functions, interfaces, types)
  const aAddedLines = parseAddedLines(aDiffs);
  const symbolPattern = /(?:export\s+)?(?:class|function|interface|type|const|enum)\s+(\w+)/;
  const aSymbols = new Set<string>();
  for (const line of aAddedLines) {
    const match = symbolPattern.exec(line);
    if (match?.[1] && match[1].length > 2) { // Skip very short identifiers
      aSymbols.add(match[1]);
    }
  }

  if (aSymbols.size === 0) return 0; // A didn't introduce new symbols

  // Check how many of B's added lines reference A's symbols
  const bAddedLines = parseAddedLines(bDiffs);
  if (bAddedLines.length === 0) return 0;

  let referencingCount = 0;
  for (const line of bAddedLines) {
    for (const sym of aSymbols) {
      if (line.includes(sym)) {
        referencingCount++;
        break; // Count each B line only once
      }
    }
  }

  return referencingCount / bAddedLines.length;
}

/**
 * Compute full work leverage analysis across an ordered list of session transcripts.
 */
export function analyzeWorkLeverage(transcripts: AgentTranscript[]): WorkLeverageResult {
  const reworkRatios: number[] = [];
  const lineSurvivalRates: number[] = [];
  const continuationIndices: number[] = [];

  for (let i = 0; i < transcripts.length; i++) {
    // Rework ratio: only for pairs (session i → session i+1)
    if (i < transcripts.length - 1) {
      reworkRatios.push(computeReworkRatio(transcripts[i]!, transcripts[i + 1]!));
    }

    // Line survival: for each session vs all later sessions
    const laterSessions = transcripts.slice(i + 1);
    lineSurvivalRates.push(computeLineSurvivalRate(transcripts[i]!, laterSessions));

    // Continuation index: for each session after the first
    if (i > 0) {
      continuationIndices.push(computeContinuationIndex(transcripts[i - 1]!, transcripts[i]!));
    }
  }

  const avg = (arr: number[]) => arr.length > 0 ? arr.reduce((a, b) => a + b, 0) / arr.length : 0;

  return {
    reworkRatios,
    avgReworkRatio: avg(reworkRatios),
    lineSurvivalRates,
    avgLineSurvivalRate: avg(lineSurvivalRates),
    continuationIndices,
    avgContinuationIndex: avg(continuationIndices),
  };
}

/**
 * Convert work leverage result to a 0-100 score.
 *
 * Composite:
 * - Low rework ratio → good (inverted: 1 - ratio)
 * - High line survival → good
 * - High continuation index → good
 *
 * Weights: rework 40%, survival 30%, continuation 30%
 */
export function workLeverageScore(result: WorkLeverageResult): number {
  const reworkScore = (1 - result.avgReworkRatio) * 100;
  const survivalScore = result.avgLineSurvivalRate * 100;
  const continuationScore = result.avgContinuationIndex * 100;

  return Math.round(
    reworkScore * 0.4 +
    survivalScore * 0.3 +
    continuationScore * 0.3,
  );
}

File: analysis/src/benchmark_analysis/dimensions/work_leverage.py
"""Work leverage metrics computed from session file changes."""
from __future__ import annotations

import re
from collections import defaultdict

from ..models import SessionTranscript


def parse_added_lines(diff: str) -> list[str]:
    """Extract added lines from a unified diff, stripping the '+' prefix."""
    return [
        line[1:].strip()
        for line in diff.split('\n')
        if line.startswith('+') and not line.startswith('+++') and line[1:].strip()
    ]


def parse_removed_lines(diff: str) -> list[str]:
    """Extract removed lines from a unified diff, stripping the '-' prefix."""
    return [
        line[1:].strip()
        for line in diff.split('\n')
        if line.startswith('-') and not line.startswith('---') and line[1:].strip()
    ]


def _is_source_file(path: str) -> bool:
    """Return True if path is a TypeScript source file in src/ or tests/."""
    return (path.startswith('src/') or path.startswith('tests/')) and path.endswith('.ts')


# Pattern for symbol declarations in TypeScript
_SYMBOL_PATTERN = re.compile(
    r'(?:export\s+)?(?:class|interface|function|const|let|type|enum)\s+(\w+)'
)


def _extract_symbol_names(lines: list[str]) -> set[str]:
    """Extract declared symbol names from added lines."""
    symbols: set[str] = set()
    for line in lines:
        for m in _SYMBOL_PATTERN.finditer(line):
            symbols.add(m.group(1))
    return symbols


def _get_session_diffs(session: SessionTranscript) -> tuple[list[str], list[str]]:
    """Get all added and removed lines from source-file changes in a session.

    Returns (added_lines, removed_lines).
    """
    added: list[str] = []
    removed: list[str] = []
    for fc in session.fileChanges:
        if not _is_source_file(fc.path):
            continue
        if fc.diff:
            added.extend(parse_added_lines(fc.diff))
            removed.extend(parse_removed_lines(fc.diff))
    return added, removed


def analyze_work_leverage(transcripts: list[SessionTranscript]) -> dict:
    """Analyze work leverage across consecutive sessions.

    For each pair of consecutive sessions (same condition+scenario+iteration),
    computes rework ratio, line survival, and continuation index.

    Returns dict with:
      - per_condition: list of dicts with aggregate metrics per condition
      - per_pair: list of dicts with per-pair details (for deeper inspection)
    """
    # Group sessions by (condition, scenario, iteration) and order by taskIndex
    groups: dict[tuple[str, str, int], list[SessionTranscript]] = defaultdict(list)

    # Determine iteration from runId — sessions share a runId within an iteration
    # Group by (condition, scenario, runId) to get sessions in same iteration
    for t in transcripts:
        # runId encodes the iteration context
        key = (t.condition, t.scenario, t.runId)
        groups[key].append(t)

    # Sort each group by taskIndex
    for key in groups:
        groups[key].sort(key=lambda t: t.taskIndex)

    per_pair: list[dict] = []
    condition_metrics: dict[str, list[dict]] = defaultdict(list)

    for (condition, scenario, run_id), sessions in groups.items():
        if len(sessions) < 2:
            continue

        for i in range(len(sessions) - 1):
            session_a = sessions[i]
            session_b = sessions[i + 1]

            added_a, _ = _get_session_diffs(session_a)
            added_b, removed_b = _get_session_diffs(session_b)

            # Rework ratio: lines B removed that A added / lines A added
            if added_a:
                added_a_set = set(added_a)
                removed_by_b_of_a = sum(1 for line in removed_b if line in added_a_set)
                rework_ratio = removed_by_b_of_a / len(added_a)
            else:
                rework_ratio = 0.0

            # Line survival: fraction of A's added lines NOT removed by later sessions
            # For simplicity, check against session B's removals only
            if added_a:
                survived = sum(1 for line in added_a if line not in set(removed_b))
                survival = survived / len(added_a)
            else:
                survival = 1.0

            # Continuation index: fraction of B's added lines that reference
            # A's new symbol names
            symbols_a = _extract_symbol_names(added_a)
            if added_b and symbols_a:
                referencing = sum(
                    1 for line in added_b
                    if any(sym in line for sym in symbols_a)
                )
                continuation = referencing / len(added_b)
            else:
                continuation = 0.0

            pair_data = {
                "condition": condition,
                "scenario": scenario,
                "run_id": run_id,
                "session_a_task": session_a.taskIndex,
                "session_b_task": session_b.taskIndex,
                "lines_added_a": len(added_a),
                "lines_added_b": len(added_b),
                "rework_ratio": round(rework_ratio, 4),
                "line_survival": round(survival, 4),
                "continuation_index": round(continuation, 4),
            }
            per_pair.append(pair_data)
            condition_metrics[condition].append(pair_data)

    # Aggregate per condition
    per_condition: list[dict] = []
    for condition in sorted(condition_metrics):
        pairs = condition_metrics[condition]
        n = len(pairs)
        avg_rework = sum(p["rework_ratio"] for p in pairs) / max(n, 1)
        avg_survival = sum(p["line_survival"] for p in pairs) / max(n, 1)
        avg_continuation = sum(p["continuation_index"] for p in pairs) / max(n, 1)
        per_condition.append({
            "condition": condition,
            "n_pairs": n,
            "avg_rework_ratio": round(avg_rework, 4),
            "avg_line_survival": round(avg_survival, 4),
            "avg_continuation_index": round(avg_continuation, 4),
        })

    return {
        "per_condition": per_condition,
        "per_pair": per_pair,
    }
